PlagiarismPlugin.inc.php

```diff
@@ -43,169 +58,292 @@ public function getDescription() {
-		if (!($folderId = $ithenticate->createFolder(
-			'Submission_' . $submission->getId(),
-			'Submission_' . $submission->getId() . ': ' . $publication->getLocalizedTitle($publication->getData('locale')),
-			$groupId,
-			true,
-			true
-		))) {
-			$this->sendErrorMessage($submission->getId(), 'Could not create folder for submission ID ' . $submission->getId() . ' on iThenticate.');
+		// $submission->setData('ithenticate_id', $submissionUuid);
+		// $submissionDao->updateObject($submission);		
+		Services::get("submission")->edit($submission, [
+			'ithenticate_id' => $submissionUuid,
+		], $request);
+		
+		// Upload submission files for successfully created submission at iThenticate's end
+		if (!$ithenticate->uploadSubmissionFile($submissionUuid, $submission)) {
+			$this->sendErrorMessage($submission->getId(), 'Could not complete the file upload at iThenticate.');
 			return false;
 		}
 
-		$submissionFiles = Services::get('submissionFile')->getMany([
-			'submissionIds' => [$submission->getId()],
-		]);
-
-		$authors = $publication->getData('authors');
-		$author = array_shift($authors);
-		foreach ($submissionFiles as $submissionFile) {
-			$file = Services::get('file')->get($submissionFile->getData('fileId'));
-			if (!$ithenticate->submitDocument(
-				$submissionFile->getLocalizedData('name'),
-				$author->getLocalizedGivenName(),
-				$author->getLocalizedFamilyName(),
-				$submissionFile->getLocalizedData('name'),
-				Services::get('file')->fs->read($file->path),
-				$folderId
-			)) {
-				$this->sendErrorMessage($submission->getId(), 'Could not submit "' . $submissionFile->getData('path') . '" to iThenticate.');
+		// If no webhook previously registered for this Site/Context, register it
+		$webhookStorable = $this->hasForcedCredentials() ? $request->getSite() : $context; /** @var Site|Context $webhookStorable */
+
+		if (!$webhookStorable->getData('ithenticate_webhook_id')) {
+			$signingSecret = \Illuminate\Support\Str::random(12);
+			
+			// http://ojs-stable-3_3_0.test/index.php/test-01/$$$call$$$/plugins/generic/plagiarism/controllers/plagiarism-webhook/handle
+			$webhookUrl = $request->getDispatcher()->url(
+                $request,
+                ROUTE_COMPONENT,
+                $context->getData('urlPath'),
+                'plugins.generic.plagiarism.controllers.PlagiarismWebhookHandler',
+                'handle'
+            );
+
+			if ($webhookId = $ithenticate->registerWebhook($signingSecret, $webhookUrl)) {
+				$webhookStorable->setData('ithenticate_webhook_signing_secret', $signingSecret);
+				$webhookStorable->setData('ithenticate_webhook_id', $webhookId);
+				$this->hasForcedCredentials()
+					? $siteDao->updateObject($webhookStorable)
+					: Application::get()->getContextDAO()->updateObject($webhookStorable);
+			} else {
+				error_log("unable to complete the iThenticate webhook registration during the submission process of ID : {$submission->getId()}");
 			}
 		}
```

Comment:
@asmecher webhook registration is failing with message like `this is not a valid URL` . we are exposing the url as such `http://ojs-stable-3_3_0.test/index.php/test-01/$$$call$$$/plugins/generic/plagiarism/controllers/plagiarism-webhook/handle` , wondering because this an offline URL(from my local machine) or because of the presence of `$$$call$$$` .

Reply:
Try using a local IP address (even 127.0.0.1) rather than the .test TLD; see if that changes the response.

Reply:
That works so I guess this is working fine .